Communications/PacketBuffer.cs

```diff
@@ -0,0 +1,373 @@
+
+                // Switch to the buffer of next prioirity
+                this.CurrentBucket++;
+                this.CurrentBucket = this.CurrentBucket >= this.NBuffers ? 0 : this.CurrentBucket;
+            }
+            return Next;
+        }
+
+        /// <summary> Get the total number of packets in the buffer. </summary>
+        public override int Count
+        {
+            get
+            {
+                int Sum = 0;
+                foreach (PacketBuffer Buffer in this.Buffers) { Sum += Buffer.Count; }
+                return Sum;
+            }
+        }
+    }
+
+    /// <summary>
+    /// This defines a generic controller.
+    /// The order of absolute priority is EMERGENT > HIGH == MEDIUM == LOW > LOWEST,
+    /// and the bandwidth allocation of HIGH : MEDIUM : LOW is 3 : 2 : 1.
+    /// </summary>
+    public class GenericController : PacketBuffer
+    {
+        public readonly QueueBuffer[] Buffers;
+        public readonly PriorityBuffer PriorityBuffer;
+        public readonly BandwidthControlBuffer BandwidthBuffer;
+
+        /// <summary> Construct a generic Controller. </summary>
+        public GenericController()
+        {
+            this.Buffers = new QueueBuffer[5];
+            for (int i = 0; i < 5; i++) { this.Buffers[i] = new QueueBuffer(); }
+
+            PacketBuffer[] BandwidthSubcontrollers = { this.Buffers[1], this.Buffers[2], this.Buffers[3] };
+            int[] BandwidthAllocation = { 3, 2, 1 };
+            this.BandwidthBuffer = new BandwidthControlBuffer(BandwidthSubcontrollers, BandwidthAllocation);
+
+            PacketBuffer[] PrioritySubcontrollers = { this.Buffers[0], this.BandwidthBuffer, this.Buffers[4] };
+            this.PriorityBuffer = new PriorityBuffer(PrioritySubcontrollers);
+        }
+
+        /// <summary> Add a packet to buffer. </summary>
+        /// 
+        /// <param name="Packet"> The packet to add. </param>
+        /// <param name="Priority"> Priority of packet. </param>
+        /// 
+        /// <exception cref="ArgumentOutOfRangeException"> If priority is out of range. </exception>
+        public override void Enqueue(Packet Packet, int Priority)
+        {
+            if (Priority < 0 || Priority >= 5) { throw new ArgumentOutOfRangeException("Priority number is out of range"); }
+            this.Buffers[Priority].Enqueue(Packet);
+        }
+
+        /// <summary> Add a packet to buffer. </summary>
+        /// 
+        /// <param name="Packet"> The packet to add. </param>
+        /// <param name="Priority"> Priority of packet. </param>
+        /// 
+        /// <exception cref="ArgumentOutOfRangeException"> If priority is out of range. </exception>
+        public void Enqueue(Packet Packet, PacketPriority Priority = PacketPriority.MEDIUM)
+        {
+            if (Priority < 0 || (int)Priority >= 5) { throw new ArgumentOutOfRangeException("Priority number is out of range"); }
```

Comment:
Would it make more sense to use say `this.Buffers.Count` instead of hard-coding `5` here?